src/bin/mir-json-translate-libs.rs

```diff
@@ -0,0 +1,692 @@
+                            "library {} should have a compiler artifact",
+                            unit.pkg_id
+                        )
+                    });
+                // Compute the patched source file path by finding the original
+                // file's relative path to its Cargo.toml, then joining that
+                // with our patched sources directory.
+                let orig_pkg_dir = artifact
+                    .manifest_path
+                    .parent()
+                    .expect("manifest path should be a file");
+                let rel_src_path =
+                    unit.target.src_path.strip_prefix(orig_pkg_dir).expect(
+                        "source file should be inside directory of Cargo.toml",
+                    );
+                let src_path = [
+                    custom_sources_dir.as_path(),
+                    Utf8Path::new(&unit.target.name),
+                    rel_src_path,
+                ]
+                .iter()
+                .collect();
+                let (linked_libs, linked_paths, cfgs, env) =
+                    match build_script_outputs.remove(&unit.pkg_id) {
+                        Some(bs) => {
+                            (bs.linked_libs, bs.linked_paths, bs.cfgs, bs.env)
+                        }
+                        None => (vec![], vec![], vec![], vec![]),
+                    };
+                CustomTarget::TargetLib(CustomTargetLib {
+                    crate_name: unit.target.name.as_str().into(),
+                    src_path,
+                    edition: unit.target.edition,
+                    linked_libs,
+                    linked_paths,
+                    // The build script's outputted cfgs does not include
+                    // features automatically computed by cargo, so we add them
+                    // in here. Technically we can get them either from the unit
+                    // graph or the artifact message, but we use the artifact
+                    // message version because it is closer to an actual
+                    // invocation of rustc. Not sure if there is actually a
+                    // difference.
+                    cfgs: artifact
+                        .features
+                        .iter()
+                        .map(|f| format!("feature=\"{f}\""))
+                        .chain(cfgs)
+                        .collect(),
+                    env,
+                })
+            }
+        },
+        dependencies: unit.dependencies,
+    };
+
+    let mut custom_graph = CustomUnitGraph {
+        units: unit_graph.units.into_iter().map(convert_unit).collect(),
+        roots: unit_graph.roots,
+    };
+
+    // Add our extra libraries to the unit graph.
+
+    let dep_core = custom_graph.get_unit_as_dep("core".into());
+    let dep_compiler_builtins =
+        custom_graph.get_unit_as_dep("compiler_builtins".into());
+    let dep_std = custom_graph.get_unit_as_dep("std".into());
```

Comment:
Could you factor out the extra libraries into a top-level constant? That way, it is easy to tell at a glance what they are, and it will be simpler to add more later if need be.

Reply:
Do you just mean the strings of the library names? Or do you mean the whole `CustomUnit`?

Reply:
I had the strings representing the library names in mind when I wrote that comment.

Reply:
Are the strings used in multiple places?  IMO there's not much value in changing `f("foo")` to `const FOO = "foo"; ...; f(FOO)`.  But if the same string is used in multiple places, then it's good to make sure they don't get out of sync.

Reply:
They are not used in multiple places.

(well, technically they also form part of the source path for each library, but I think those should probably be kept separate.)